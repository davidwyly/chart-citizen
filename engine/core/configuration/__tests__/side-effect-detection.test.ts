/**
 * Side Effect Detection Test Suite
 * ================================
 * 
 * This comprehensive test suite is designed to detect any unintended side effects
 * from the rendering path pipeline refactor. It establishes baseline behavior
 * and validates that the refactored system produces identical results.
 * 
 * Key Areas Tested:
 * - Orbital mechanics calculations remain identical
 * - Camera positioning is consistent
 * - View mode switching behavior is preserved
 * - Object visibility logic is unchanged
 * - Performance characteristics are maintained
 * - Memory usage patterns are stable
 * - No unintended state mutations occur
 */

import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import * as THREE from 'three';

// Mock the system data for consistent testing
const createMockSystemData = () => ({
  objects: [
    {
      id: 'sol',
      name: 'Sol',
      classification: 'star',
      properties: { radius: 695700000, mass: 1.989e30 },
      orbit: null
    },
    {
      id: 'earth',
      name: 'Earth',
      classification: 'planet',
      properties: { radius: 6371000, mass: 5.972e24 },
      orbit: { semi_major_axis: 1.0, eccentricity: 0.0167, parent: 'sol' }
    },
    {
      id: 'luna',
      name: 'Luna',
      classification: 'moon',
      properties: { radius: 1737000, mass: 7.342e22 },
      orbit: { semi_major_axis: 0.00257, eccentricity: 0.0549, parent: 'earth' }
    },
    {
      id: 'jupiter',
      name: 'Jupiter',
      classification: 'planet',
      properties: { radius: 69911000, mass: 1.898e27 },
      orbit: { semi_major_axis: 5.2, eccentricity: 0.0489, parent: 'sol' }
    }
  ]
});

// Create mock object refs map
const createMockObjectRefsMap = () => {
  const map = new Map<string, THREE.Object3D>();
  
  // Sol
  const sol = new THREE.Mesh();
  sol.position.set(0, 0, 0);
  sol.scale.set(2.0, 2.0, 2.0);
  sol.userData = { id: 'sol', name: 'Sol' };
  map.set('sol', sol);
  
  // Earth
  const earth = new THREE.Mesh();
  earth.position.set(50, 0, 0);
  earth.scale.set(1.0, 1.0, 1.0);
  earth.userData = { id: 'earth', name: 'Earth', orbitRadius: 50 };
  map.set('earth', earth);
  
  // Luna
  const luna = new THREE.Mesh();
  luna.position.set(52, 0, 0);
  luna.scale.set(0.3, 0.3, 0.3);
  luna.userData = { id: 'luna', name: 'Luna', orbitRadius: 2 };
  map.set('luna', luna);
  
  // Jupiter
  const jupiter = new THREE.Mesh();
  jupiter.position.set(260, 0, 0);
  jupiter.scale.set(4.0, 4.0, 4.0);
  jupiter.userData = { id: 'jupiter', name: 'Jupiter', orbitRadius: 260 };
  map.set('jupiter', jupiter);
  
  return { current: map };
};

describe('Side Effect Detection', () => {
  let mockSystemData: any;
  let mockObjectRefsMap: any;
  let originalConsoleLog: any;
  let originalConsoleWarn: any;
  let originalConsoleError: any;
  
  beforeEach(() => {
    mockSystemData = createMockSystemData();
    mockObjectRefsMap = createMockObjectRefsMap();
    
    // Suppress console output during tests unless there are actual errors
    originalConsoleLog = console.log;
    originalConsoleWarn = console.warn;
    originalConsoleError = console.error;
    console.log = vi.fn();
    console.warn = vi.fn();
    console.error = vi.fn();
  });
  
  afterEach(() => {
    // Restore console
    console.log = originalConsoleLog;
    console.warn = originalConsoleWarn;
    console.error = originalConsoleError;
  });
  
  describe('Orbital Mechanics Baseline', () => {
    it('should calculate consistent visual radii for all view modes', async () => {
      // Import orbital mechanics calculator
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      const viewModes = ['explorational', 'navigational', 'profile', 'scientific'];
      const results: Record<string, any> = {};
      
      // Calculate for each view mode
      for (const viewMode of viewModes) {
        try {
          const result = calculateSystemOrbitalMechanics(mockSystemData.objects, viewMode as any);
          results[viewMode] = result;
        } catch (error) {
          // Store error for analysis
          results[viewMode] = { error: error.message };
        }
      }
      
      // Validate that all view modes produce results
      for (const viewMode of viewModes) {
        expect(results[viewMode]).toBeDefined();
        if (results[viewMode].error) {
          console.error(`Error in ${viewMode}:`, results[viewMode].error);
        }
        expect(results[viewMode]).not.toHaveProperty('error');
      }
      
      // Store baseline results for future comparison
      expect(results).toMatchSnapshot();
    });
    
    it('should maintain consistent object hierarchy', async () => {
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      const result = calculateSystemOrbitalMechanics(mockSystemData.objects, 'explorational');
      
      // Verify parent-child relationships are preserved
      const earthResult = result.get('earth');
      const lunaResult = result.get('luna');
      const solResult = result.get('sol');
      const jupiterResult = result.get('jupiter');
      
      expect(earthResult).toBeDefined();
      expect(lunaResult).toBeDefined();
      expect(solResult).toBeDefined();
      expect(jupiterResult).toBeDefined();
      
      // Sol should be largest
      expect(solResult.visualRadius).toBeGreaterThan(jupiterResult.visualRadius);
      expect(solResult.visualRadius).toBeGreaterThan(earthResult.visualRadius);
      
      // Jupiter should be larger than Earth (allow for slight variations due to scaling)
      expect(jupiterResult.visualRadius).toBeGreaterThanOrEqual(earthResult.visualRadius * 0.9);
      
      // Earth should be larger than Luna
      expect(earthResult.visualRadius).toBeGreaterThan(lunaResult.visualRadius);
      
      // Luna should have an orbit distance greater than Earth's radius
      if (lunaResult.orbitDistance !== undefined) {
        expect(lunaResult.orbitDistance).toBeGreaterThan(earthResult.visualRadius);
      }
    });
    
    it('should prevent collisions in all view modes', async () => {
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      const viewModes = ['explorational', 'navigational', 'profile', 'scientific'];
      
      for (const viewMode of viewModes) {
        const result = calculateSystemOrbitalMechanics(mockSystemData.objects, viewMode as any);
        
        // Check for planet-planet collisions
        const earthResult = result.get('earth');
        const jupiterResult = result.get('jupiter');
        
        if (earthResult?.orbitDistance && jupiterResult?.orbitDistance) {
          const earthOuterEdge = earthResult.orbitDistance + earthResult.visualRadius;
          const jupiterInnerEdge = jupiterResult.orbitDistance - jupiterResult.visualRadius;
          
          expect(earthOuterEdge).toBeLessThan(jupiterInnerEdge);
        }
        
        // Check for moon-planet collisions
        const lunaResult = result.get('luna');
        if (lunaResult?.orbitDistance && earthResult) {
          const lunaInnerEdge = lunaResult.orbitDistance - lunaResult.visualRadius;
          expect(lunaInnerEdge).toBeGreaterThan(earthResult.visualRadius);
        }
      }
    });
  });
  
  describe('Camera Controller Baseline', () => {
    it('should calculate consistent camera positions', () => {
      // Test camera distance calculations
      const testCases = [
        { visualSize: 1.0, expectedMin: 2.5, expectedOptimal: 4.0, expectedMax: 15.0 },
        { visualSize: 2.0, expectedMin: 5.0, expectedOptimal: 8.0, expectedMax: 30.0 },
        { visualSize: 0.5, expectedMin: 1.25, expectedOptimal: 2.0, expectedMax: 7.5 },
      ];
      
      for (const testCase of testCases) {
        const { visualSize, expectedMin, expectedOptimal, expectedMax } = testCase;
        
        // These are the current hardcoded multipliers from unified-camera-controller
        const CONSISTENT_RADIUS_MULTIPLIER = 4.0;
        const CONSISTENT_MIN_MULTIPLIER = 2.5;
        const CONSISTENT_MAX_MULTIPLIER = 15.0;
        
        const optimalDistance = visualSize * CONSISTENT_RADIUS_MULTIPLIER;
        const minDistance = visualSize * CONSISTENT_MIN_MULTIPLIER;
        const maxDistance = visualSize * CONSISTENT_MAX_MULTIPLIER;
        
        expect(optimalDistance).toBeCloseTo(expectedOptimal, 2);
        expect(minDistance).toBeCloseTo(expectedMin, 2);
        expect(maxDistance).toBeCloseTo(expectedMax, 2);
      }
    });
    
    it('should maintain view mode specific elevation angles', () => {
      // These are the current elevation angles from view mode configurations
      const expectedAngles = {
        explorational: 30,
        navigational: 35,
        profile: 22.5,
        scientific: 15,
      };
      
      // For now, just verify the baseline values are documented
      // After refactor, we'll verify the configuration system provides these
      for (const [mode, angle] of Object.entries(expectedAngles)) {
        expect(angle).toBeGreaterThan(0);
        expect(angle).toBeLessThan(90);
      }
    });
    
    it('should handle profile view layout calculations consistently', () => {
      // Test the profile view fake offset logic
      const testCases = [
        { visualSize: 1.0, expectedOffset: 3.0 },
        { visualSize: 2.0, expectedOffset: 6.0 },
        { visualSize: 0.5, expectedOffset: 1.5 },
      ];
      
      for (const testCase of testCases) {
        const { visualSize, expectedOffset } = testCase;
        
        // Current fake offset multiplier from camera controller
        const fakeOffset = visualSize * 3;
        
        expect(fakeOffset).toBeCloseTo(expectedOffset, 2);
      }
    });
  });
  
  describe('Memory and Performance Baseline', () => {
    it('should not create memory leaks during calculations', async () => {
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      // Force garbage collection if available
      if (global.gc) {
        global.gc();
      }
      
      const initialMemory = process.memoryUsage();
      
      // Perform calculations multiple times
      for (let i = 0; i < 100; i++) {
        calculateSystemOrbitalMechanics(mockSystemData.objects, 'explorational');
        calculateSystemOrbitalMechanics(mockSystemData.objects, 'navigational');
        calculateSystemOrbitalMechanics(mockSystemData.objects, 'profile');
        calculateSystemOrbitalMechanics(mockSystemData.objects, 'scientific');
      }
      
      // Force garbage collection again if available
      if (global.gc) {
        global.gc();
      }
      
      const finalMemory = process.memoryUsage();
      
      // Memory usage should not have increased significantly
      const heapUsedIncrease = finalMemory.heapUsed - initialMemory.heapUsed;
      const heapUsedIncreasePercent = (heapUsedIncrease / initialMemory.heapUsed) * 100;
      
      // Allow up to 50% increase in heap usage (very generous threshold)
      expect(heapUsedIncreasePercent).toBeLessThan(50);
    });
    
    it('should maintain calculation performance', async () => {
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      const iterations = 100;
      const start = performance.now();
      
      for (let i = 0; i < iterations; i++) {
        calculateSystemOrbitalMechanics(mockSystemData.objects, 'explorational');
      }
      
      const elapsed = performance.now() - start;
      const avgTime = elapsed / iterations;
      
      // Each calculation should complete in under 10ms (generous threshold)
      expect(avgTime).toBeLessThan(10);
    });
  });
  
  describe('State Mutation Detection', () => {
    it('should not mutate input system data', async () => {
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      const originalData = JSON.stringify(mockSystemData);
      
      calculateSystemOrbitalMechanics(mockSystemData.objects, 'explorational');
      calculateSystemOrbitalMechanics(mockSystemData.objects, 'navigational');
      calculateSystemOrbitalMechanics(mockSystemData.objects, 'profile');
      calculateSystemOrbitalMechanics(mockSystemData.objects, 'scientific');
      
      const finalData = JSON.stringify(mockSystemData);
      
      expect(finalData).toBe(originalData);
    });
    
    it('should not mutate Three.js objects during camera calculations', () => {
      const originalPositions = new Map();
      const originalScales = new Map();
      
      // Record original states
      for (const [id, obj] of mockObjectRefsMap.current) {
        originalPositions.set(id, obj.position.clone());
        originalScales.set(id, obj.scale.clone());
      }
      
      // Simulate camera controller operations
      for (const [id, obj] of mockObjectRefsMap.current) {
        // Get world position (this should not mutate the object)
        const worldPos = new THREE.Vector3();
        obj.getWorldPosition(worldPos);
        
        // Calculate distances (this should not mutate the object)
        const distance = worldPos.length();
        expect(distance).toBeGreaterThanOrEqual(0);
      }
      
      // Verify no mutations occurred
      for (const [id, obj] of mockObjectRefsMap.current) {
        const originalPos = originalPositions.get(id);
        const originalScale = originalScales.get(id);
        
        expect(obj.position.equals(originalPos)).toBe(true);
        expect(obj.scale.equals(originalScale)).toBe(true);
      }
    });
  });
  
  describe('View Mode Switching Consistency', () => {
    it('should produce deterministic results for repeated calculations', async () => {
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      const viewModes = ['explorational', 'navigational', 'profile', 'scientific'];
      
      for (const viewMode of viewModes) {
        const result1 = calculateSystemOrbitalMechanics(mockSystemData.objects, viewMode as any);
        const result2 = calculateSystemOrbitalMechanics(mockSystemData.objects, viewMode as any);
        
        // Results should be identical
        expect(result1.size).toBe(result2.size);
        
        for (const [id, data1] of result1) {
          const data2 = result2.get(id);
          expect(data2).toBeDefined();
          expect(data1.visualRadius).toBeCloseTo(data2!.visualRadius, 10);
          
          if (data1.orbitDistance !== undefined && data2!.orbitDistance !== undefined) {
            expect(data1.orbitDistance).toBeCloseTo(data2!.orbitDistance, 10);
          }
        }
      }
    });
    
    it('should handle edge cases consistently', async () => {
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      // Test with empty objects array
      const emptyResult = calculateSystemOrbitalMechanics([], 'explorational');
      expect(emptyResult.size).toBe(0);
      
      // Test with single object
      const singleObject = [mockSystemData.objects[0]];
      const singleResult = calculateSystemOrbitalMechanics(singleObject, 'explorational');
      expect(singleResult.size).toBe(1);
      
      // Test with objects that have no orbit data
      const noOrbitObject = [{
        id: 'test',
        name: 'Test',
        classification: 'star',
        properties: { radius: 1000, mass: 1000 },
        orbit: null
      }];
      const noOrbitResult = calculateSystemOrbitalMechanics(noOrbitObject, 'explorational');
      expect(noOrbitResult.size).toBe(1);
    });
  });
  
  describe('Integration Points', () => {
    it('should maintain compatibility with existing Three.js scene structure', () => {
      // Test that our mock structure matches expected patterns
      for (const [id, obj] of mockObjectRefsMap.current) {
        expect(obj).toBeInstanceOf(THREE.Object3D);
        expect(obj.userData).toHaveProperty('id');
        expect(obj.userData).toHaveProperty('name');
        expect(obj.position).toBeInstanceOf(THREE.Vector3);
        expect(obj.scale).toBeInstanceOf(THREE.Vector3);
      }
    });
    
    it('should handle missing or invalid object references gracefully', () => {
      const incompleteRefsMap = { current: new Map() };
      incompleteRefsMap.current.set('earth', mockObjectRefsMap.current.get('earth')!);
      // Missing other objects
      
      // This should not throw an error
      expect(() => {
        for (const [id, obj] of incompleteRefsMap.current) {
          const worldPos = new THREE.Vector3();
          obj.getWorldPosition(worldPos);
        }
      }).not.toThrow();
    });
  });
  
  describe('Error Handling Baseline', () => {
    it('should handle malformed system data gracefully', async () => {
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      const malformedData = [
        {
          id: 'malformed',
          name: 'Malformed',
          classification: 'planet',
          properties: null, // Missing properties object
          orbit: null
        }
      ];
      
      // Current behavior: throws on malformed data (this is the baseline we're establishing)
      expect(() => {
        calculateSystemOrbitalMechanics(malformedData as any, 'explorational');
      }).toThrow();
    });
    
    it('should handle invalid view modes gracefully', async () => {
      const { calculateSystemOrbitalMechanics } = await import('/home/dwyly/code/chart-citizen/engine/core/pipeline');
      
      // Should not throw, should fall back to default behavior
      expect(() => {
        calculateSystemOrbitalMechanics(mockSystemData.objects, 'invalid' as any);
      }).not.toThrow();
    });
  });
});

/**
 * Regression Test Data Storage
 * ============================
 * 
 * This section stores the baseline calculations that should remain
 * consistent throughout the refactor. Any changes to these values
 * indicate potential side effects that need investigation.
 */
export const BASELINE_TEST_DATA = {
  systemData: createMockSystemData(),
  expectedHierarchy: {
    sol: { minRadius: 2.0, isRoot: true },
    jupiter: { minRadius: 1.0, parent: 'sol' },
    earth: { minRadius: 0.5, parent: 'sol' },
    luna: { minRadius: 0.1, parent: 'earth' }
  },
  expectedDistanceRatios: {
    // Jupiter should be roughly 5x farther than Earth
    jupiterToEarth: { min: 4.0, max: 6.0 },
    // Luna should be close to Earth
    lunaToEarth: { min: 0.001, max: 0.1 }
  },
  performanceThresholds: {
    maxCalculationTime: 10, // ms
    maxMemoryIncrease: 50, // percent
    maxCacheSize: 1000 // number of entries
  }
};

/**
 * Side Effect Detection Utilities
 * ===============================
 * 
 * Utility functions for detecting and reporting side effects
 * during the refactor process.
 */
export class SideEffectDetector {
  private static baselines: Map<string, any> = new Map();
  
  static recordBaseline(key: string, value: any): void {
    this.baselines.set(key, structuredClone(value));
  }
  
  static compareWithBaseline(key: string, currentValue: any): {
    matches: boolean;
    differences: string[];
  } {
    const baseline = this.baselines.get(key);
    if (!baseline) {
      return { matches: false, differences: ['No baseline recorded'] };
    }
    
    const differences: string[] = [];
    
    try {
      expect(currentValue).toEqual(baseline);
      return { matches: true, differences: [] };
    } catch (error) {
      differences.push(error.message);
      return { matches: false, differences };
    }
  }
  
  static generateReport(): string {
    const report = ['=== Side Effect Detection Report ==='];
    report.push(`Baselines recorded: ${this.baselines.size}`);
    
    for (const [key] of this.baselines) {
      report.push(`- ${key}`);
    }
    
    return report.join('\n');
  }
}