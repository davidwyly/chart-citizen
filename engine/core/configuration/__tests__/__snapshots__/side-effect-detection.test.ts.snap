// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Side Effect Detection > Orbital Mechanics Baseline > should calculate consistent visual radii for all view modes 1`] = `
{
  "explorational": Map {
    "sol" => {
      "visualRadius": 1,
    },
    "earth" => {
      "orbitDistance": 50,
      "visualRadius": 0.8179249725317846,
    },
    "jupiter" => {
      "orbitDistance": 260,
      "visualRadius": 0.8,
    },
    "luna" => {
      "orbitDistance": 6.25,
      "visualRadius": 0.6816041437764873,
    },
  },
  "navigational": Map {
    "sol" => {
      "visualRadius": 15801.57589075498,
    },
    "earth" => {
      "orbitDistance": 1806.3,
      "visualRadius": 1200,
    },
    "jupiter" => {
      "orbitDistance": 19408.67589075498,
      "visualRadius": 13167.979908962485,
    },
    "luna" => {
      "orbitDistance": 3600,
      "visualRadius": 0.6,
    },
  },
  "profile": Map {
    "sol" => {
      "visualRadius": 1.7999999999999998,
    },
    "earth" => {
      "orbitDistance": 9.75,
      "visualRadius": 1.7999999999999998,
    },
    "jupiter" => {
      "orbitDistance": 19.05,
      "visualRadius": 1.5,
    },
    "luna" => {
      "orbitDistance": 5.25,
      "visualRadius": 1.5,
    },
  },
  "scientific": Map {
    "sol" => {
      "visualRadius": 7.800000000000001,
    },
    "earth" => {
      "orbitDistance": 80,
      "visualRadius": 0.07142992669253989,
    },
    "jupiter" => {
      "orbitDistance": 416,
      "visualRadius": 0.7838231996550238,
    },
    "luna" => {
      "orbitDistance": 0.20559999999999998,
      "visualRadius": 0.01947477360931436,
    },
  },
}
`;
